whats-that-typeclass-functor/article.md

```diff
@@ -0,0 +1,467 @@
+
+The implementation of `fmap` is data-type-specific. So to use it on a data type, that data type needs to have an implementation for it.
+
+In Haskell, this means that the type needs to have an instance of the Functor typeclass.
+
+## The Functor typeclass
+
+What’s Functor? Let’s ask GHCi.
+
+```haskell
+> :info Functor
+type Functor :: (* -> *) -> Constraint
+class Functor f where
+  fmap :: (a -> b) -> f a -> f b
+  (<$) :: a -> f b -> f a
+  {-# MINIMAL fmap #-}
+```
+
+Functor in Haskell is a [typeclass](https://serokell.io/blog/haskell-typeclasses) that describes two common functions — `fmap` and `(<$)`.
+
+The kind of Functor is `(* -> *) -> Constraint`, where `* -> *` is the kind of data type which could be a Functor. You can read about kinds and the `* -> *` kind especially in our [previous article](https://serokell.io/blog/whats-that-typeclass-foldable?nocache=#generalization-of-foldr-and-foldl).
+
+Functor provides methods for structure-preserving transformations.
+
+You’re already familiar with `fmap`. However, just any `fmap` is not enough for a correct Functor instance. Some Haskell typeclasses have laws — conditions that instances have to meet. Only this guarantees that the instance is correct.
+
+Functor, in particular, needs `fmap` to adhere to the following laws.
+
+---
+
+### Laws of Functor
+
+1. **Identity**
+
+    `fmap id == id` — applying `id` function to the wrapped value changes nothing.
+
+    Example:
+
+    `fmap id (Just 1) == id (Just 1) == Just 1`
+
+2. **Composition**
+
+    `fmap f . fmap g == fmap (f . g)` — applying `fmap`s sequentially is the same as applying `fmap` with the composition of functions.
+
+    Example:
+
+    `fmap (+1) (fmap (*2) (Just 1)) == fmap ((+1) . (*2)) (Just 1) == Just 3`
+
+
+---
+
+`fmap` is all you need to define a `Functor` instance. However, the `(<$)` operator is worth looking at too.
+
+### The `(<$)` operator
+
+Take a look at the type signature of `(<$)`:
+
+```haskell
+(<$) :: a -> f b -> f a
+```
+
+The operator takes a value of type `a`, a value of type `b` packed in a functor `f` – `f b` – and produces a functor `f a`. Basically, the operator packs the first argument’s value in the context of the second argument, throwing away the second value.
+
+Now, let’s try to guess the definition of `(<$)` using the provided description and the following examples.
+
+```haskell
+-- replace an `Int` value
+-- inside `Maybe Int`
+-- with another `Int`
+> 1 <$ Just 2
+Just 1
+> 1 <$ Nothing
+Nothing
+
+-- replace an `Int` value
+-- inside `Maybe Int`
+-- with a `String`
+> "abc" <$ Just 123
+Just "abc"
+> "abc" <$ Nothing
+Nothing
+
+-- replace each element
+-- of the list `[Int]`
+-- with an `Int`
+> 1 <$ []
+[]
+> 1 <$ [0]
+[1]
+> 1 <$ [1..5]
+[1,1,1,1,1]
+
+-- replace each element
+-- of the list `[Int]`
+-- with a `Char`
+-- Note: the resulted type
+-- is `String` as `type String = [Char]`
```

Comment:
I am not sure if I got it correctly

@serokellcao, do you propose to change this example in the following way?
```haskell
-- replace each element
-- of the list `[Int]`
-- with a `Char`
> '1' <$ []
[]
> '1' <$ [0]
['1']
> '1' <$ [1..5]
['1', '1', '1', '1', '1']
```
Or do you mean that it's better to remove this example completely, because any mentioning `type String = [Char]`  is undesired? 🤔 

Anyway, it seems strange to me to hide this misfeature from readers. The very first usage of GHCi would promote `String = [Char]`, so it looks unfair to avoid this in the article.

Reply:
Yeah, it would be better not to touch charlists. 

Given that the example would work with regular 1, I would prefer that.